include/circt/Dialect/Calyx/CalyxPrimitives.td

```diff
@@ -116,6 +116,74 @@ def MemoryOp : CalyxPrimitive<"memory", []> {
   }];
 }
 
+def SeqMemoryOp : CalyxPrimitive<"seq_mem", []> {
+  let summary = "Defines a memory with sequential read";
+  let description = [{
+    The "calyx.seq_mem" op defines a memory with sequential reads. Memories can 
+    have any number of dimensions, as specified by the length of the `$sizes` and
+    `$addrSizes` arrays. The `$addrSizes` specify the bitwidth of each dimension's
+    address, and should be wide enough to address the range of the corresponding
+    dimension in `$sizes`. The `$width` attribute dictates the width of a single
+    element.
+
+    See https://capra.cs.cornell.edu/docs/calyx/libraries/core.html#memories for
+    more information.
+
+    ```mlir
+      // A 1-dimensional, 32-bit memory with size dimension 1. Equivalent representation in the native compiler:
+      // `m1 = seq_mem_d1(32, 1, 1)`
+      %m1.addr0, %m1.write_data, %m1.write_en, %m1.write_done, %m1.clk, %m1.read_data, %m1.read_en, %m1.read_done = calyx.memory @m1 <[1] x 32> [1] : i1, i32, i1, i1, i32, i1
+
+      // A 2-dimensional, 8-bit memory with size dimensions 64 x 64. Equivalent representation in the native compiler:
+      // `m2 = seq_mem_d2(8, 64, 64, 6, 6)`
+      %m2.addr0, %m2.addr1, %m2.write_data, %m2.write_en, %m2.write_done, %m2.clk, %m2.read_data, %m2.read_en, %m2.read_done = calyx.memory @m2 <[64, 64] x 8> [6, 6] : i6, i6, i8, i1, i1, i8, i1
+    ```
+  }];
+
+  let arguments = (ins
+    SymbolNameAttr:$sym_name,
+    I64Attr:$width,
+    ArrayAttr:$sizes,
+    ArrayAttr:$addrSizes
+  );
+
+  let results = (outs
+    Variadic<AnySignlessInteger>:$results
```

Comment:
Really not a big fan of the long results type list (`: i1, i32, i1, i1, i32, i1` when most (all?) of those are known based on the memory type.
Could change the results to
> `let results = (outs Variadic<AnySignlessInteger>:$addrs, AnySignlessInteger:$writeData, I1:$wrEn, I1:$wrDone, I1:$clk, AnySignlessInteger:$rdData, I1:$rdEn, I1:$rdDone)>`

 + possibly having to setting the `AttrSizedResultSegments` attribute. This would also be much more canonical, in that all of the custom getters that you've implemented will be autogenerated.

Reply:
I agree, but I am just following the method used on existing Calyx ops for now. This op functions very similarly to the existing `calyx.memory` op. Maybe I can merge this PR and then work on another PR at a later date to bring all of the Calyx ops inline with the suggestions you make?

Reply:
Works for me - a general cleanup in this aspect would be nice!